File: app/src/main/java/com/agera/hometools/login/LoginActivity.java
package com.agera.hometools.login;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.util.Pair;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import com.agera.hometools.MainActivity;
import com.agera.hometools.R;
import com.google.android.agera.BaseObservable;
import com.google.android.agera.Repositories;
import com.google.android.agera.Repository;
import com.google.android.agera.Result;
import com.google.android.agera.Updatable;
import com.google.android.agera.net.HttpResponse;

import java.util.concurrent.atomic.AtomicBoolean;

public class LoginActivity extends Activity implements Updatable {

    private EditText mEt_tel = null;
    private EditText mEt_password = null;
    private Button mBtn_login = null;
    private OnClickListenerObservable mOb = null;
    private Repository<Result<HttpResponse>> mRep = null;
    private String tel = null;
    private String password = null;
    private AtomicBoolean activeOnce = new AtomicBoolean(false);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        initView();
        initEvents();
    }

    private void initEvents() {
        mOb = new OnClickListenerObservable();
        mBtn_login.setOnClickListener(mOb);
        mRep = Repositories.repositoryWithInitialValue(Result.<HttpResponse>absent())
                .observe(mOb)
                .onUpdatesPerLoop()
                .check(o -> activeOnce.getAndSet(true))
                .orSkip()
                .getFrom(() -> {
                    return tel = mEt_tel == null ? null : mEt_tel.getText() == null ? null : mEt_tel.getText().toString().trim();
                })
                .check(s -> LoginFunctionsImp.instance().checkTel(mEt_tel).apply(s))
                .orSkip()
                .getFrom(() -> {
                    return password = mEt_password == null ? null : mEt_password.getText() == null ? null : mEt_password.getText().toString().trim();
                })
                .check(s -> LoginFunctionsImp.instance().checkPassword(mEt_password).apply(s))
                .orSkip()
                .thenTransform(s -> LoginFunctionsImp.instance().login(null, mBtn_login).apply(Pair.create(tel, password)))
                .notifyIf((response1, response2) -> LoginFunctionsImp.instance().checkLogin(mBtn_login).apply(response2))
                .compile();
        activeOnce.set(false);
        mRep.addUpdatable(this);
    }

    private void initView() {
        mEt_tel = (EditText) findViewById(R.id.et_tel);
        mEt_password = (EditText) findViewById(R.id.et_password);
        mBtn_login = (Button) findViewById(R.id.btn_Login);
    }

    @Override
    public void update() {
        Log.e("---", "----logon success");
        startActivity(new Intent(this, MainActivity.class));
    }

    class OnClickListenerObservable extends BaseObservable implements View.OnClickListener {
        @Override
        public void onClick(View v) {
            dispatchUpdate();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mRep.removeUpdatable(this);
    }
}

File: app/src/main/java/com/agera/hometools/login/LoginActivity.kt
package com.agera.hometools.login

import android.content.Context
import android.content.Intent
import android.graphics.Paint
import android.os.Bundle
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.TextView
import com.agera.hometools.BaseActivity
import com.agera.hometools.MainActivity
import com.agera.hometools.MyApp
import com.agera.hometools.R
import com.agera.hometools.common.ClickObservable
import com.agera.hometools.core.TaskDriver
import com.agera.hometools.push.PushImp
import com.agera.hometools.utils.CommonUtils
import com.google.android.agera.Repositories
import com.google.android.agera.Repository
import com.google.android.agera.Result
import com.google.android.agera.Updatable

/**
 * Created by Agera on 2017/11/24.
 */
class LoginActivity : BaseActivity(), Updatable {

    var mEt_tel: EditText? = null
    var mEt_password: EditText? = null
    var activeOnce: Result<String> = Result.failure()
    var tel: String = ""
    var password: String = ""
    var mRep: Repository<Result<String>>? = null
    var loginObservable: ClickObservable? = null
    var tv_register: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        initViews()
        initEvents()
    }

    private fun initEvents() {
        loginObservable = ClickObservable()
        findViewById(R.id.btn_Login).setOnClickListener { view ->
            activeOnce = Result.success("start repository")
            (MyApp.instance().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).hideSoftInputFromWindow(view.windowToken, 0)
            loginObservable!!.onClick(view)
        }
        tv_register?.setOnClickListener {
            startActivity(Intent(this, RegisterActivity::class.java))
        }
        mRep = Repositories.repositoryWithInitialValue(Result.absent<String>())
                .observe(loginObservable)
                .onUpdatesPerLoop()
                .attemptGetFrom { activeOnce }
                .orSkip()
                .attemptGetFrom { LoginImp.instance().checkTel(mEt_tel) }
                .orSkip()
                .attemptTransform {
                    tel = it
                    LoginImp.instance().checkPassword(mEt_password)
                }
                .orSkip()
                .goTo(TaskDriver.instance().mCore)
                .typedResult(String::class.java)
                .thenTransform {
                    password = it
                    CommonUtils.instance().showShortMessage(mEt_tel!!, "正在登录...")
                    LoginImp.instance().login(tel, password)
                }
                .notifyIf { _, v2 ->
                    if (v2.failed())
                        CommonUtils.instance().showShortMessage(mEt_tel!!, "登陆失败...")
                    v2.succeeded()
                }
                .compile()

        mRep?.let { it.addUpdatable(this) }
    }

    private fun initViews() {
        mEt_tel = findViewById(R.id.et_tel) as EditText
        mEt_password = findViewById(R.id.et_password) as EditText
        tv_register = findViewById(R.id.tv_register) as TextView
        tv_register!!.paintFlags = Paint.UNDERLINE_TEXT_FLAG
    }

    override fun onDestroy() {
        super.onDestroy()
        mRep?.let {
            it.removeUpdatable(this)
        }
    }

    override fun update() {
        //auto remember account
        CommonUtils.instance().saveAccountInfo(tel,password)
        MyApp.instance().userName = tel
        //refresh push
        PushImp.instance().setPushAccount(this, tel)
        //start main page
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }

}